pkg/services/blockchain/eth/resolver.go

```diff
@@ -148,9 +245,206 @@ func (r *Resolver) Resolve(
 		}
 	}
 
+	signature := ""
+	if opts.Signature != "" {
+		if r.walletKey == "" {
+			return services.IdentityState{},
+				errors.New("no wallet key found for generating signature")
+		}
+		primaryType := services.IdentityStateType
+		if opts.GistRoot != nil {
+			primaryType = services.GlobalStateType
+		}
+		signature, err = r.signTypedData(primaryType, did, identityState)
+		if err != nil {
+			return services.IdentityState{}, err
+		}
+	}
+
+	identityState.Signature = signature
+
 	return identityState, err
 }
 
+func (r *Resolver) VerifyState(
+	primaryType services.PrimaryType,
+	identityState services.IdentityState,
+	did w3c.DID,
+) (bool, error) {
+	walletAddress, err := r.GetWalletAddress()
+	if err != nil {
+		return false, err
+	}
+
+	typedData, err := r.TypedData(primaryType, did, identityState, walletAddress)
+	if err != nil {
+		return false, err
+	}
+
+	authData := AuthData{TypedData: typedData, Signature: identityState.Signature, Address: walletAddress}
+	return r.verifyTypedData(authData)
+}
+
+func TimeStampFn() string {
+	timestamp := strconv.FormatInt(time.Now().UTC().Unix(), 10)
+	return timestamp
+}
+
+func (r *Resolver) TypedData(primaryType services.PrimaryType, did w3c.DID, identityState services.IdentityState, walletAddress string) (apitypes.TypedData, error) {
+	id, err := core.IDFromDID(did)
+	if err != nil {
+		return apitypes.TypedData{},
+			fmt.Errorf("invalid did format for did '%s': %v", did, err)
+	}
+	ID := id.BigInt().String()
+	idType := fmt.Sprintf("0x%02X%02X", id.Type()[0], id.Type()[1])
+
+	apiTypes := apitypes.Types{}
+	message := apitypes.TypedDataMessage{}
+	primaryTypeString := ""
+	timestamp := TimeStamp()
+
+	switch primaryType {
+	case services.IdentityStateType:
+		state := "0"
+		replacedAtTimestamp := "0"
+
+		if identityState.StateInfo != nil {
```

Comment:
I guess if primaryType is IdentityState and identityState.StateInfo is nil this function even must not be called. Same for GistInfo.

Reply:
Yes, checked at the beginning of the function and throwing error in case the info is nil